consensus/istanbul/core/backlog.go

```diff
@@ -103,114 +116,181 @@ func (c *core) storeBacklog(msg *istanbul.Message, src istanbul.Validator) {
+		return
+	}
+
+	for !backlogForSeq.Empty() {
+		m := backlogForSeq.PopItem()
+		msg := m.(*istanbul.Message)
+		if cb != nil {
+			_, src := c.valSet.GetByAddress(msg.Address)
+			if src != nil {
+				cb(msg, src)
+			}
+		}
+		c.backlogCountByVal[msg.Address]--
+		c.backlogTotal--
+	}
+	delete(c.backlogBySeq, seq)
 }
 
 func (c *core) processBacklog() {
+
 	c.backlogsMu.Lock()
 	defer c.backlogsMu.Unlock()
 
-	for src, backlog := range c.backlogs {
-		if backlog == nil {
-			continue
-		}
+	for _, seq := range c.getSortedBacklogSeqs() {
 
-		logger := c.logger.New("from", src, "state", c.state, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "processBacklog")
-		isFuture := false
-
-		// We stop processing if
-		//   1. backlog is empty
-		//   2. The first message in queue is a future message
-		for !(backlog.Empty() || isFuture) {
-			m, prio := backlog.Pop()
-			msg := m.(*istanbul.Message)
-			var view *istanbul.View
-			switch msg.Code {
-			case istanbul.MsgPreprepare:
-				var m *istanbul.Preprepare
-				err := msg.Decode(&m)
-				if err == nil {
-					view = m.View
-				}
-			case istanbul.MsgPrepare:
-				fallthrough
-			case istanbul.MsgCommit:
-				var sub *istanbul.Subject
-				err := msg.Decode(&sub)
-				if err == nil {
-					view = sub.View
+		logger := c.logger.New("state", c.state, "seq", seq)
+
+		if seq < c.currentView().Sequence.Uint64() {
+			// Earlier sequence. Prune all messages.
+			c.drainBacklogForSeq(seq, nil)
+		} else if seq == c.currentView().Sequence.Uint64() {
+			// Current sequence. Process all in order.
+			c.drainBacklogForSeq(seq, func(msg *istanbul.Message, src istanbul.Validator) {
+				var view *istanbul.View
+				switch msg.Code {
+				case istanbul.MsgPreprepare:
+					var m *istanbul.Preprepare
+					err := msg.Decode(&m)
+					if err == nil {
+						view = m.View
+					}
+				case istanbul.MsgPrepare:
+					fallthrough
+				case istanbul.MsgCommit:
+					var sub *istanbul.Subject
+					err := msg.Decode(&sub)
+					if err == nil {
+						view = sub.View
+					}
+				case istanbul.MsgRoundChange:
+					var rc *istanbul.RoundChange
+					err := msg.Decode(&rc)
+					if err == nil {
+						view = rc.View
+					}
 				}
-			case istanbul.MsgRoundChange:
-				var rc *istanbul.RoundChange
-				err := msg.Decode(&rc)
-				if err == nil {
-					view = rc.View
+				if view == nil {
+					logger.Debug("Nil view", "msg", msg)
+					// continue
+					return
 				}
-			}
-			if view == nil {
-				logger.Debug("Nil view", "msg", msg)
-				continue
-			}
-			// Push back if it's a future message
-			err := c.checkMessage(msg.Code, view)
-			if err != nil {
-				if err == errFutureMessage {
-					logger.Trace("Stop processing backlog", "msg", msg)
-					backlog.Push(msg, prio)
-					isFuture = true
-					break
+				err := c.checkMessage(msg.Code, view)
+				if err != nil {
+					if err == errFutureMessage {
+						// TODO(asa): Why is this unexpected? It could be for a future round...
```

Comment:
How would this happen? I thought `seq` is the current sequence, and all the messages in this callback should be from this sequence? I would go as far to say that this should be an error level message?

Reply:
What if they're from a future round?

Reply:
We are draining for the current sequence? https://github.com/celo-org/celo-blockchain/pull/517/files/6f0aed30022a89bb06009932c12969fc7a521cd8#diff-3bcb34249a8e59b2eca3deb619ec7238R233